src/core/cluster.py

```diff
@@ -0,0 +1,183 @@
+        self._servers_data = {}
+
+        self.peer_app_data = DataPeerData(
+            self.model, relation_name=PEER, additional_secret_fields=PEER_APP_SECRETS
+        )
+        self.peer_unit_data = DataPeerUnitData(
+            self.model, relation_name=PEER, additional_secret_fields=PEER_UNIT_SECRETS
+        )
+        self.client_requires_data = OpenSearchRequiresData(
+            self.model,
+            relation_name=OPENSEARCH_REL_NAME,
+            index=DASHBOARD_INDEX,
+            extra_user_roles="kibanaserver",
+        )
+
+    # --- RAW RELATION ---
+
+    @property
+    def peer_relation(self) -> Relation | None:
+        """The cluster peer relation."""
+        return self.model.get_relation(PEER)
+
+    @property
+    def opensearch_relation(self) -> Optional[Relation]:
+        """The Opensearch Server relation."""
+        return self.model.get_relation(OPENSEARCH_REL_NAME)
+
+    # --- CORE COMPONENTS---
+
+    @property
+    def unit_server(self) -> ODServer:
+        """The server state of the current running Unit."""
+        return ODServer(
+            relation=self.peer_relation,
+            data_interface=self.peer_unit_data,
+            component=self.model.unit,
+            substrate=self.substrate,
+        )
+
+    @property
+    def peer_units_data(self) -> Dict[Unit, DataPeerOtherUnitData]:
+        """The cluster peer relation."""
+        if not self.peer_relation or not self.peer_relation.units:
+            return {}
+
+        for unit in self.peer_relation.units:
+            if unit not in self._servers_data:
+                self._servers_data[unit] = DataPeerOtherUnitData(
+                    model=self.model, unit=unit, relation_name=PEER
+                )
+        return self._servers_data
+
+    @property
+    def cluster(self) -> ODCluster:
+        """The cluster state of the current running App."""
+        return ODCluster(
+            relation=self.peer_relation,
+            data_interface=self.peer_app_data,
+            component=self.model.app,
+            substrate=self.substrate,
+        )
+
+    @property
+    def servers(self) -> Set[ODServer]:
+        """Grabs all servers in the current peer relation, including the running unit server.
+
+        Returns:
+            Set of ODServers in the current peer relation, including the running unit server.
+        """
+        if not self.peer_relation:
+            return set()
+
+        servers = set()
+        for unit, data_interface in self.peer_units_data.items():
+            servers.add(
+                ODServer(
+                    relation=self.peer_relation,
+                    data_interface=data_interface,
+                    component=unit,
+                    substrate=self.substrate,
+                )
+            )
+        servers.add(self.unit_server)
+
+        return servers
+
+    @property
+    def opensearch_server(self) -> Optional[OpensearchServer]:
```

Comment:
nit: no particular opinions tbh, but if you're using `A | B`, then I'd recommend keeping the same pattern here

Reply:
Actually there were a bunch of other outliers, all updated.

Reply:
Done.